lib/arpv4/arpv4.ml

```diff
@@ -94,68 +87,53 @@ module Make (Ethif : V1_LWT.ETHIF) (Clock : V1.CLOCK) (Time : V1_LWT.TIME) = str
     | Not_found ->
       Hashtbl.replace t.cache ip (Confirmed (expire, mac))
 
-  (* Input handler for an ARP packet, registered through attach() *)
-  let rec input t frame =
-    let open Arpv4_wire in
-    MProf.Trace.label "arpv4.input";
-    match get_arp_op frame with
-    |1 -> (* Request *)
-      (* Received ARP request, check if we can satisfy it from
-         our own IPv4 list *)
-      let req_ipv4 = Ipaddr.V4.of_int32 (get_arp_tpa frame) in
-      (* printf "ARP: who-has %s?\n%!" (Ipaddr.V4.to_string req_ipv4); *)
-      if List.mem req_ipv4 t.bound_ips then begin
-        Log.info (fun f -> f "ARP responding to: who-has %a?" Ipaddr.V4.pp_hum req_ipv4);
-        (* We own this IP, so reply with our MAC *)
-        let sha = Ethif.mac t.ethif in
-        let tha = Macaddr.of_bytes_exn (copy_arp_sha frame) in
-        let spa = Ipaddr.V4.of_int32 (get_arp_tpa frame) in (* the requested address *)
-        let tpa = Ipaddr.V4.of_int32 (get_arp_spa frame) in (* the requesting host IPv4 *)
-        output t { op=`Reply; sha; tha; spa; tpa }
-      end else Lwt.return_unit
-    |2 -> (* Reply *)
-      let spa = Ipaddr.V4.of_int32 (get_arp_spa frame) in
-      let sha = Macaddr.of_bytes_exn (copy_arp_sha frame) in
-      Log.info (fun f -> f "ARP: updating %a -> %s"
-        Ipaddr.V4.pp_hum spa (Macaddr.to_string sha));
-      (* If we have pending entry, notify the waiters that answer is ready *)
-      notify t spa sha;
-      Lwt.return_unit
-    |n ->
-      Log.info (fun f -> f "ARP: Unknown message %d ignored" n);
-      Lwt.return_unit
-
-  and output t arp =
-    let open Arpv4_wire in
+  let output t arp =
+    let open Arpv4_parse in
     (* Obtain a buffer to write into *)
-    let buf = Io_page.to_cstruct (Io_page.get 1) in
+    let buf = Cstruct.create (Ethif_wire.sizeof_ethernet + Arpv4_wire.sizeof_arp) in
     (* Write the ARP packet *)
-    let dmac = Macaddr.to_bytes arp.tha in
-    let smac = Macaddr.to_bytes arp.sha in
-    let spa = Ipaddr.V4.to_int32 arp.spa in
-    let tpa = Ipaddr.V4.to_int32 arp.tpa in
-    let op =
+    match Arpv4_print.print_arpv4_header
+      ~buf:(Cstruct.shift buf Ethif_wire.sizeof_ethernet)
+      ~src_ip:arp.spa ~dst_ip:arp.tpa ~src_mac:arp.sha ~dst_mac:arp.tha
+      ~op:arp.op with
+    | Error s -> Log.info (fun f -> f "Failed to print Arpv4 header: %s" s);
```

Comment:
Shouldn't this be more serious than an info-level log message? Since we allocate the buffer just above, presumably at the right size, we should probably `failwith s` here to indicate a bug.